### ISSUE TITLE
Is this even legal?

### ISSUE THREAD

**Issue by AlexanderMaxRanabel** [WangTingZheng/mcp940#2]

**Comment by samuelh2005** [comment]
No. This is just a reupload of the MCP code. The usage of the code reproduced by MCP is controlled by their [LICESNSE](https://github.com/WangTingZheng/mcp940/blob/master/LICENSE.txt) which is included in this repository, and it states:
```
You are NOT allowed to:
- Use MCP to do anything that violated Mojangs terms of use for Minecraft.
- Release Minecraft versions or modifications that allow you to play without having bought Minecraft from Mojang.
- Release modified or unmodified versions of MCP anywhere.
- Use any of MCPs scripts, tools or data files without explicit written permission.
- Make money with anything based on MCP (excluding Minecraft mods created by using MCP).
- Use MCP to create clients that are used for griefing or exploiting server bugs.
- Release the decompiled source code of Minecraft in any way.
```

**Comment by thasave14** [comment]
but nobody cares here so

**Comment by GiwbyAlbatross** [comment]
Then what is the contents of src/minecraft/net/minecraft/ ? It seems to be minecraft’s source code and the repo description is ‘source code of Minecraft 1.12’.

**Comment by samuelh2005** [comment]
> Then what is the contents of src/minecraft/net/minecraft/ ? It seems to be minecraft’s source code and the repo description is ‘source code of Minecraft 1.12’.

@GiwbyAlbatross The contents are actually the output from MCP. Mojang's official code is obfuscated and not public. However tools like MCP take Mojang's obfuscated code and applies human readable names (deobfusucation).

However the logic in the code remains the same as Mojang's official code.

**Comment by GiwbyAlbatross** [comment]
I really don’t want this to get taken down, but I think under the mcp license you are not allowed to “Release the decompiled source code of Minecraft in any way.” Which this repo kinda is doing. Having said that, this is a fantastic resource for learning Minecraft modding, Java, or game development full stop.

**Comment by TomLucidor** [comment]
If anyone here has the balls to "refactor in Rust" (including making it mod-compatible) then it probably would have different limitations.

**Comment by samuelh2005** [comment]
> If anyone here has the balls to "refactor in Rust" (including making it mod-compatible) then it probably would have different limitations.

@TomLucidor Making a mod compatible version of Minecraft in Rust is impossible becuase:
1. Mods are made in Java because the game is made in Java
2. Mod loaders (like Forge) inject custom code into Mincraft at runtime (or at installation)
3. Mods expect classes and interfaces to exist where they do normally (big problem - Rust doesn't have classes or interfaces)
4. Java compiles into Java specfic bytecode executed by a JVM. Rust is compiled into machine code that the CPU executes.

You are right that it would have different limitations. The rust code would be enterily new and the person making it would get to decide the license. They just need to make sure they don't use any Mojang's offical assets.

The closest to mod compatiblitly that a Rust Minecraft would actually work is datapack compatiblitly or Bedrock Edition addons compatiblitly.

**Comment by TomLucidor** [comment]
@ajh123 what is the best way to "FOSS clean room" with a JVM compatible language then?

**Comment by samuelh2005** [comment]
> [@ajh123](https://github.com/ajh123) what is the best way to "FOSS clean room" with a JVM compatible language then?

It is possible to create a Minecraft clone in Java. However the legal problem lies in copying the existing structure of Minecraft. So whilst it could be possible to create a Minecraft clone that is compatible with mods techically it would be illegal as you need to make sure your Minecraft follows the excat same code as Mojang's.

The most legal and FOSS way to make a Minecraft in a JVM compatible language would be to make a Minecraft clone from scratch, and support Bedrock addons or Java datapacks. You could also support Spigot / Paper plugins as their API is independent from Mojang.

**Comment by GiwbyAlbatross** [comment]
I found a way (in theory) to deobfuscate the names of the audio (and other things in `jars/assets/objects`) but would that be a copyright issue if I extract and release the audio?

**Comment by samuelh2005** [comment]
> I found a way (in theory) to deobfuscate the names of the audio (and other things in `jars/assets/objects`) but would that be a copyright issue if I extract and release the audio?

@GiwbyAlbatross Correct, that would be a copyright issue. That is basically the same thing as buying a music CD and putting a different label on it then selling it.

One way to get around this is to automatically download the Minecraft jar file from Mojang when your game starts, so that you can extract the music from it - This is the same process that Fabric and Forge does *except they are not dealing with music but the entire code*.

**Comment by GiwbyAlbatross** [comment]
I have taken the data and resources from 1.21.5 and put them [here](https://www.github.com/GiwbyAlbatross/vanilla). I specified that they don’t belong to me, and so on. Am I allowed to do that? I figure that if this repo can release the whole source code, I can release an up-to-date version of the default data pack, right?

**Comment by samuelh2005** [comment]
> I have taken the data and resources from 1.21.5 and put them [here](https://www.github.com/GiwbyAlbatross/vanilla). I specified that they don’t belong to me, and so on. Am I allowed to do that? I figure that if this repo can release the whole source code, I can release an up-to-date version of the default data pack, right?

@GiwbyAlbatross 

Leagly speaking Mojang still owns all the code and textures. However it seems like Mojang's enforcement team do not look at GitHub often, so whilst you can publish code and textures to GItHub it is a gray area.

The [Minecraft EULA](https://www.minecraft.net/en-us/eula) says:

> However, you must not distribute anything we've made unless we specifically agree to it. By "distribute anything we've made" what we mean is:
> - give copies of our game software or content to anyone else;
> - make commercial use of anything we've made;
> - try to make money from anything we've made; or
> - let other people get access to anything we've made in a way that is unfair or unreasonable.

and also says:

> The [Microsoft Services Agreement](https://www.microsoft.com/servicesagreement) says “Your Content remains Your Content”, and > that applies to Minecraft. We don't own the original stuff that you create. We will however own things that are copies (or substantial copies) or derivatives of our property and creations - but if you create original things, they aren't ours. So, as an example:
>
> - a single Minecraft block (including its textures and its “look and feel”) - we own that;
> -  your creation of a Gothic Cathedral with a rollercoaster running through it - we don't own that.

These lines in the EULA confirm you cannot distruibute code or textures. If you want to get around this you can use the Faithfull resource pack - which is allmost similar to the default - *but make sure to follow their license too!*

**Comment by GiwbyAlbatross** [comment]
Turns out it’s been done a handful of times across GitHub with version history all the way back to 18w...
